### PR TITLE
Lighting rework

### PR DESCRIPTION
**Pull request by Cyclonit**

### CODE REVIEW

File: src/main/java/cubicchunks/world/IOpacityIndex.java [hunk 1]
```diff
@@ -47,6 +47,16 @@
...
 
 	/**
+	 * Returns true if the block at the given position is occluded by a known non-opaque block further up.
+	 *
+	 * @param localX local block x-coordinate (0..15)
+	 * @param blockY global block y-coordinate
+	 * @param localZ local block z-coordinate (0..15)
+	 * @return true if there exists a known non-opaque block above the given y-coordinate
+	 */
+	boolean isOccluded(int localX, int blockY, int localZ);
```
Comment (Barteks2x) [inline src/main/java/cubicchunks/world/IOpacityIndex.java]: Why add yet another method to interface if it  can be already done using everything that already is there with no performance difference?

Comment (Cyclonit) [inline src/main/java/cubicchunks/world/IOpacityIndex.java]: Determining if a given block-position is occluded is contextually the OpacityIndex's job. If a different kind of LightManager happens to use the OpacityIndex it'd likely have to implement this method if it weren't already there.

File: src/main/java/cubicchunks/lighting/FirstLightProcessor.java [hunk 2]
```diff
@@ -60,155 +56,198 @@
...
 
-		//cache min/max Y, generating them may be expensive
+		// Cache min/max Y, generating them may be expensive
 		int[][] minBlockYArr = new int[16][16];
 		int[][] maxBlockYArr = new int[16][16];
 
-		int minBlockX = cubeToMinBlock(generatedCube.getX());
-		int maxBlockX = cubeToMaxBlock(generatedCube.getX());
+		int minBlockX = Coords.cubeToMinBlock(cube.getX());
+		int maxBlockX = Coords.cubeToMaxBlock(cube.getX());
```
Comment (Cyclonit) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: Renaming generatedCube to cube makes it conform to the overall standard in this class. This class itself does not impose a restriction on diffuseSkylight being called for generatedCubes only.
I'll put the static import back in, personal preference of mine not to use them, as they blur the line of which methods are implemented in which class.

File: src/main/java/cubicchunks/lighting/FirstLightProcessor.java [hunk 3]
```diff
@@ -60,155 +56,198 @@
...
-				new Int2ObjectOpenCustomHashMap<>(10, 0.75f, CUBE_Y_HASH);
+		Int2ObjectMap<FastCubeBlockAccess> blockAccessMap = new Int2ObjectOpenCustomHashMap<>(10, 0.75f, CUBE_Y_HASH);
 
-		Column column = generatedCube.getColumn();
+		Column column = cube.getColumn();
 		for (int blockX = minBlockX; blockX <= maxBlockX; blockX++) {
 			for (int blockZ = minBlockZ; blockZ <= maxBlockZ; blockZ++) {
-				pos.setPos(blockX, pos.getY(), blockZ);
+
+				this.mutablePos.setPos(blockX, this.mutablePos.getY(), blockZ);
```
Comment (Cyclonit) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: But you loose the knowledge that it is a field and not a local variable. Having it this way shows possible race-conditions immediately.

File: src/main/java/cubicchunks/lighting/FirstLightProcessor.java [hunk 4]
```diff
@@ -60,155 +56,198 @@
...
 				}
 
-				Iterable<Cube> cubes = column.getLoadedCubes(blockToCube(maxBlockY), blockToCube(minBlockY));
-				int topBlockY = getHeightmapValue(column, blockToLocal(blockX), blockToLocal(blockZ));
-				//going top-down
-				for(Cube cube : cubes) {
-					//generatedCube must be updated
-					//the logic in canStopUpdating doesn't apply to it
-					if(cube != generatedCube && canStopUpdating(cube, pos, topBlockY)) {
-						break;
```
Comment (Cyclonit) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: Reintroduced

File: src/main/java/cubicchunks/lighting/FirstLightProcessor.java [hunk 5]
```diff
@@ -60,155 +56,198 @@
...
-					}
-					//if FirstLight isn't done in tha cube yet (and it's not the current cube) - t will be done later
-					//so skip it there
-					if(cube != generatedCube && !cube.isInitialLightingDone()) {
-						continue;
```
Comment (Cyclonit) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: Reintroduced

**Pull request merged**